### PR TITLE
[feat] v3 - dicom microscopy extension and mode

### PR DESCRIPTION
**Pull request by md-prog**
### Context

[Dicom Microscopy Viewer](https://github.com/ImagingDataCommons/dicom-microscopy-viewer) integration for OHIF v3.

### Changes & Results

- New extension `@ohif/extension-dicom-microscopy` is added
- New mode `@ohif/mode-microscopy` is added

### TODO

We should merge this after the new DICOMLoader gets merged which will fix the overwrite for the web worker.
(There was some conflicts between CSWIL and DMV when exporting webworker files.)

#### PR

- [x] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [x] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

- [] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [x] OS: Windows 10
- [x] Node version: 16.15.0
- [x] Browser: Chrome Version 111.0.5563.65 (Official Build) (64-bit)

[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch

### Acknowledgements

- [DICOM Microscopy Viewer](https://github.com/ImagingDataCommons/dicom-microscopy-viewer) is a Vanilla JS library for web-based visualization of DICOM VL Whole Slide Microscopy Image datasets and derived information.
- [SLIM Viewer](https://github.com/imagingdatacommons/slim) is a single-page application for interactive visualization and annotation of digital whole slide microscopy images and derived image analysis results in standard DICOM format. The application is based on the dicom-microscopy-viewer JavaScript library and runs fully client side without any custom server components.
- Huge thanks to [Freie Universität Berlin](https://www.fu-berlin.de/)

### CODE REVIEW

File: platform/ui/src/assets/icons/tool-circle.svg
Comment (sedghi) [inline platform/ui/src/assets/icons/tool-circle.svg:1]: where are these 4 icons being used?

File: .webpack/webpack.base.js
Comment (sedghi) [inline .webpack/webpack.base.js:156]: Do we really need this? how come our other wasm codecs for loading work without this?

File: platform/core/src/services/DicomMetadataStore/createStudyMetadata.js
Comment (sedghi) [inline platform/core/src/services/DicomMetadataStore/createStudyMetadata.js:30]: good catch

File: platform/viewer/.webpack/webpack.pwa.js
Comment (sedghi) [inline platform/viewer/.webpack/webpack.pwa.js:136]: this will break the main viewer via CSWIL, but please add a todo in the PR description that we should merge this after the new DICOMLoader gets merged which will fix the overwrite for the webworker

File: platform/viewer/public/config/default.js
Comment (sedghi) [inline platform/viewer/public/config/default.js:32]: revert

File: platform/viewer/src/routes/Local/Local.tsx
Comment (sedghi) [inline platform/viewer/src/routes/Local/Local.tsx:85]: please add a Todo that:
 we are forcing to go to `microscopy` mode which is not ideal, we should make the local drag and drop navigate to the worklist and there user can select microscopy mode

Comment (sedghi) [reply platform/viewer/src/routes/Local/Local.tsx:85]: How about something similar to this

```js
let studies = await filesToStudies(acceptedFiles, dataSource)
let queryParams = ""

if (microscopyExtensionLoaded) {
  const smStudies = studies.filter((id) => {
    const study = DicomMetadataStore.getStudy(id)
    return (
      study.series.findIndex(
        (s) => s.Modality === "SM" || s.instances[0].Modality === "SM"
      ) >= 0
    )
  })

  if (smStudies.length > 0) {
    studies = smStudies
    queryParams = "microscopy"
  }
}

const studyInstanceUIDs = studies
  .map((id) => `StudyInstanceUIDs=${id}`)
  .join("&")
const url = `/${queryParams}/dicomlocal?${studyInstanceUIDs}`

navigate(url)
```

File: modes/microscopy/public/assets/icons/tool-circle.svg
Comment (sedghi) [inline modes/microscopy/public/assets/icons/tool-circle.svg:1]: where do we inject these icons that are inside your mode?

File: modes/microscopy/README.md
Comment (sedghi) [inline modes/microscopy/README.md:4]: please give attribution to SLIIM Viewer and DMV

File: extensions/dicom-microscopy/README.md
Comment (sedghi) [inline extensions/dicom-microscopy/README.md:4]: please add proper attribution to SLIM and DMV

File: extensions/dicom-microscopy/src/utils/Publisher.js
Comment (sedghi) [inline extensions/dicom-microscopy/src/utils/Publisher.js:4]: we have a pubsub service, please use that. Look at CornerstoneViewportService as an example.

File: extensions/dicom-microscopy/src/tools/microscopyManager.js
Comment (sedghi) [inline extensions/dicom-microscopy/src/tools/microscopyManager.js:22]: this service (manager) should get registered to servicesManager. Look at out other extension's services for that. For instance see `ToolGroupService` there is a REGISTRATION that should be added to your manager here (I recommend calling it service). Then you can grab its instance from servicesManager.services throughout the app

File: extensions/dicom-microscopy/src/tools/microscopyManager.js
Comment (sedghi) [inline extensions/dicom-microscopy/src/tools/microscopyManager.js:153]: in the context of OHIF, viewer is OHIF, and we have viewports. Do you meen viewport here?

File: extensions/dicom-microscopy/src/tools/microscopyManager.js
Comment (sedghi) [inline extensions/dicom-microscopy/src/tools/microscopyManager.js:233]: should this be `addViewport` to match ohif language?

File: extensions/dicom-microscopy/src/tools/microscopyManager.js
Comment (sedghi) [inline extensions/dicom-microscopy/src/tools/microscopyManager.js:332]: use ternery operator

File: extensions/dicom-microscopy/src/tools/viewerManager.js
Comment (sedghi) [inline extensions/dicom-microscopy/src/tools/viewerManager.js:23]: same for this, it makes sense to have them registered in services manager of ohif instead of re-deising the architecture

File: extensions/dicom-microscopy/src/utils/RoiAnnotation.js
Comment (sedghi) [inline extensions/dicom-microscopy/src/utils/RoiAnnotation.js:14]: same here, move all services inside a folder called services and use OHIF services manager

File: extensions/dicom-microscopy/src/utils/loadSR.js
Comment (sedghi) [inline extensions/dicom-microscopy/src/utils/loadSR.js:19]: remove

File: extensions/dicom-microscopy/src/DicomMicroscopyViewport.tsx
Comment (sedghi) [inline extensions/dicom-microscopy/src/DicomMicroscopyViewport.tsx:71]: how about 

```js
const symbols = Object.getOwnPropertySymbols(this.viewer);
const _drawingSource = symbols.find(p => p.description === 'drawingSource');
const _pyramid = symbols.find(p => p.description === 'pyramid');
const _map = symbols.find(p => p.description === 'map');
const _affine = symbols.find(p => p.description === 'affine');
```

File: extensions/dicom-microscopy/src/DicomMicroscopyViewport.tsx
Comment (sedghi) [inline extensions/dicom-microscopy/src/DicomMicroscopyViewport.tsx:76]: return early to reduce indentation

```js
if (!feature){
  return null
}

// rest of logic
```

File: extensions/dicom-microscopy/src/components/MicroscopyLabelDialog/MicroscopyLabelDialog.styl
Comment (sedghi) [inline extensions/dicom-microscopy/src/components/MicroscopyLabelDialog/MicroscopyLabelDialog.styl:1]: can we jsut remove these two .style files since there is no .style in v3, it was in v2. Maybe work it with tailwindcss

**Review by sedghi — COMMENTED** [review]
great work indeed. My major comment is how managers are created in isolation instead of getting created similar to other services that we have

File: platform/ui/src/assets/icons/tool-circle.svg
Comment (md-prog) [reply platform/ui/src/assets/icons/tool-circle.svg:1]: They are used by annotation tools for dicom microscopy.
Removing unused icons in "mode"

File: modes/microscopy/public/assets/icons/tool-circle.svg
Comment (md-prog) [reply modes/microscopy/public/assets/icons/tool-circle.svg:1]: Removed those icons, as we have added the same icons into '/platform/viewer'

**Comment by md-prog** [comment]
@sedghi Thank you for the review. The patches have been pushed.

**Comment by fedorov** [comment]
Can someone please explain, and ideally provide a sample object, how the annotations are saved and what is the resulting DICOM object?

**Comment by sedghi** [comment]
@fedorov As discussed in the email, we hided the export until we get the feedback from you guys regarding the SR

**Review by sedghi — APPROVED** [review]
Looks great. This is a great contribution, thanks a lot!

**Pull request merged**
